src/phongo_log.c

```diff
@@ -0,0 +1,193 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "bson/bson.h"
+#include "mongoc/mongoc.h"
+
+#include <stdio.h>
+
+#include <php.h>
+#include <ext/date/php_date.h>
+#include <main/php_open_temporary_file.h>
+#include <Zend/zend_exceptions.h>
+#include <Zend/zend_interfaces.h>
+#include <Zend/zend_operators.h>
+
+#include "php_phongo.h"
+#include "phongo_error.h"
+#include "phongo_log.h"
+
+ZEND_EXTERN_MODULE_GLOBALS(mongodb)
+
+static void phongo_log_to_stream(FILE* stream, mongoc_log_level_t level, const char* domain, const char* message)
+{
+	struct timeval tv;
+	time_t         t;
+	zend_long      tu;
+	zend_string*   dt;
+
+	bson_gettimeofday(&tv);
+	t  = tv.tv_sec;
+	tu = tv.tv_usec;
+
+	dt = php_format_date((char*) ZEND_STRL("Y-m-d\\TH:i:s"), t, 0);
+
+	fprintf(stream, "[%s.%06" PHONGO_LONG_FORMAT "+00:00] %10s: %-8s> %s\n", ZSTR_VAL(dt), tu, domain, mongoc_log_level_str(level), message);
+	fflush(stream);
+	efree(dt);
+}
+
+/* Dispatch a log message to all loggers in a HashTable. The caller is
+ * responsible for ensuring that loggers implement the correct interface. */
+static void phongo_log_dispatch(mongoc_log_level_t level, const char* domain, const char* message)
+{
+	zval* logger;
+	zval func_name;
+	zval args[3];
+
+	ZVAL_STRING(&func_name, "log");
+	ZVAL_LONG(&args[0], level);
+	ZVAL_STRING(&args[1], domain);
+	ZVAL_STRING(&args[2], message);
+
+	ZEND_HASH_FOREACH_VAL_IND(MONGODB_G(loggers), logger)
+	{
+		zval retval;
+
+		if (EG(exception)) {
+			break;
+		}
+
+		call_user_function(NULL, logger, &func_name, &retval, 3, args);
+		zval_ptr_dtor(&retval);
+	}
+	ZEND_HASH_FOREACH_END();
+
+	zval_ptr_dtor(&func_name);
+	zval_ptr_dtor(&args[0]);
+	zval_ptr_dtor(&args[1]);
+	zval_ptr_dtor(&args[2]);
+}
+
+static void phongo_log_handler(mongoc_log_level_t level, const char* domain, const char* message, void* user_data)
+{
+	if (MONGODB_G(debug_fd)) {
+		phongo_log_to_stream(MONGODB_G(debug_fd), level, domain, message);
+	}
+
+	if (MONGODB_G(loggers) && zend_hash_num_elements(MONGODB_G(loggers)) > 0) {
+		phongo_log_dispatch(level, domain, message);
+	}
+}
+
+/* Sets or unsets our libmongoc handler according to whether logging is enabled
+ * (i.e. there is a stream or the loggers HashTable is not empty). This should
+ * be called each time after updating the stream or logger HashTable. */
+static void phongo_log_sync_handler(void)
+{
+	if (MONGODB_G(debug_fd) || (MONGODB_G(loggers) && zend_hash_num_elements(MONGODB_G(loggers)) > 0)) {
+		mongoc_log_trace_enable();
```

Comment:
This is the optimization I was referring to above. It's responsible for setting a `gLogTrace` global that libmongoc uses to early-return in `mongoc_log()`.

Note: this is different than the optimization proposed in [CDRIVER-4697](https://jira.mongodb.org/browse/CDRIVER-4697), although related.